_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+<script src="main.js"></script>
+</body>
+</html>
+```
+
+Create the fragment `navigation.html` at `src/main/resources/templates/fragments`:
+
+```html
+<!-- src/main/resources/templates/fragments/navigation.html -->
+<nav class="navbar navbar-expand-lg navbar-dark bg-dark">
+    <div class="container">
+        <a class="navbar-brand" href="#">Navbar</a>
+        <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+            <span class="navbar-toggler-icon"></span>
+        </button>
+        <div class="collapse navbar-collapse" id="navbarSupportedContent">
+            <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+                <li class="nav-item">
+                    <a class="nav-link active" aria-current="page" href="#">Home</a>
+                </li>
+                <li class="nav-item">
+                    <a class="nav-link" href="#">Link</a>
+                </li>
+            </ul>
+            <ul class="navbar-nav d-flex mb-2 mb-lg-0">
+                <li class="nav-item dropdown">
+                    <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
+                        [[${username}]]
+                    </a>
+                    <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
+                        <li><a class="dropdown-item" th:href="@{/profile}">Profile</a></li>
+                        <li>
+                            <form method="post" th:action="@{/logout}">
+                                <input type="hidden" th:name="${_csrf.parameterName}" th:value="${_csrf.token}"/>
+                                <button id="logout" class="dropdown-item" type="submit">Logout</button>
+                            </form>
+                        </li>
+                    </ul>
+                </li>
+            </ul>
+        </div>
+    </div>
+</nav>
+```
+
+Create the `HomeController.java` in the `com.example.demo.web` package:
+
+```java
+// src/main/java/com/example/demo/web/HomeController.java
+package com.example.demo.web;
+
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
+public class HomeController {
+
+    @GetMapping("/")
+    public String index(@AuthenticationPrincipal OidcUser oidcUser, Model model) {
+        model.addAttribute("username", oidcUser.getEmail());
+        return "index";
+    }
+}
+```
+
+Add a `profile.html` template:
```

Comment:
I dont think we need seperate profile page, use the template I mentioned above so that on authentication the page automatically shows profile. Uupdate the mapping for `/` to return `ModelAndView` as in https://github.com/oktadev/auth0-spring-boot-passkeys-demo/blob/main/src/main/java/com/example/demo/HomeController.java

Reply:
Updated according to the example